m3u8/version_matching.py

```diff
@@ -0,0 +1,43 @@
+from typing import Callable, List
+
+from m3u8 import protocol
+from m3u8.version_matching_rules import *
+
+
+class VersionMatchingError(Exception):
+    def __init__(self, lineno, line):
+        self.lineno = lineno
+        self.line = line
+
+    def __str__(self):
+        return f"Version matching error at line {self.lineno}: {self.line}"
+
+
+def get_version(file_lines: List[str]):
+    for line in file_lines:
+        if line.startswith(protocol.ext_x_version):
+            version = line.split(":")[1]
+            return float(version)
+
+    return None
+
+
+def valid_in_all_rules(line_number: int, line: str, version: float):
+    rules: List[Callable[[str, float], bool]] = [
+        valid_iv_in_EXT_X_KEY,
+        valid_floating_point_EXTINF,
+        valid_EXT_X_BYTERANGE_or_EXT_X_I_FRAMES_ONLY,
+    ]
+
+    for rule in rules:
+        if not rule(line, version):
+            raise VersionMatchingError(line_number, line)
```

Comment:
I like the whole idea. Thanks for the work! I'd suggest adding why it failed so developers can know how to fix it.

Something in the lines of:

```python
    rules: List[Callable[[str, float], bool]] = [
        valid_iv_in_EXT_X_KEY,
        valid_floating_point_EXTINF,
        valid_EXT_X_BYTERANGE_or_EXT_X_I_FRAMES_ONLY,
    ]

    for rule in rules:
        if not rule(line, version):
            raise VersionMatchingError(line_number, line, f"failed to run the checker {rule.__name__}")
 # or even using a more descriptive help; encoding rules as a dict, k -> v, k being the string describing the check and the value being the function/callback
# maybe add the fields reason and how_to_fix attached to each validation, I dunno
```

Also, maybe we should have a list of errors wrapped into a final exception so the user can know a list of what to fix instead of running it multiple times.

```python
errors = []
# ...
if len(errors) > 0:
  raise ManifestInvalid(... , errors)
```

Reply:
Thank you! I agree with you, I'll work on that!

Reply:
Done! Could you please take a look on that?